### ISSUE TITLE
implement median filter for trajectories

### ISSUE THREAD

**Issue by espinielli** [euctrl-pru/trrrj#10]
provide an API for a median filtering feature.
This is useful to smooth "spiky" trajectories such as initial output from [RT](https://github.com/euctrl-pru/reftrj)

It could be implemented out of a [median filter](https://en.wikipedia.org/wiki/Median_filter) such as in [`stats::runmed`](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/runmed.html) (see also [`zoo::rollmean`](https://www.rdocumentation.org/packages/zoo/versions/1.8-6/topics/rollmean) or [`signal::medfilt1`](https://www.rdocumentation.org/packages/signal/versions/0.7-6/topics/medfilt1))

An initial implementation in [R/smooth.R](https://github.com/euctrl-pru/trrrj/commit/78835ee189e7ab3a8a2015ee09a0d3e0bf480637#diff-dfcbf3d1f52cc6977b6132dfe7f4c8b3) is from [`pracma::movavg`](https://www.rdocumentation.org/packages/pracma/versions/1.9.9/topics/movavg).

See also @xoolive's implementation of [`filter`](https://github.com/xoolive/traffic/blob/master/traffic/core/flight.py#L640) in his [`trajectory`](https://github.com/xoolive/traffic/) Python library.
It caters for:
1. forward and backward NA filling via Pandas's Dataframe `x.bfill().ffill()`.
   In R we could use [`tidyr::fill()`](https://tidyr.tidyverse.org/reference/fill.html)
2. applying a custom filter, by default a median one via [`scipy.signal.medfilt`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.medfilt.html).
   Noteworthy is the quality criteria to accept the filtering, see [these lines](https://github.com/xoolive/traffic/blob/14de7a23bbe56e5304cbe05758a299e100aa44dd/traffic/core/flight.py#L733-L734)
   ```python
            # Decision to accept/reject for all data points in the time series
            new_data.loc[df.sq_eps > df.sigma, feat] = None
   ```
   Where `sq_eps` is the average of the squared differences, i.e. `sq_eps = (y - y_median) ** 2`,
   and `sigma` is
   ```python
      sigma = np.sqrt(filt(sq_eps, kernel_size))
   ```
   Value not matching the quality criteria will be replaced by a NaN (`None`) and
   the fill strategy in 1. will kick in to fill the holes.

**Comment by xoolive** [comment]
As you carefully noted in 2. the implementation is a bit different than a medfilt, but built around it. 

The issue with data from OpenSky database is more about faulty values stored in base than about data needing to be filtered (the aircraft does already some kind of filtering on the data before emitting it). So we try to find irrelevant values instead of averaging everything. It usually works better in cascade.

Good default values for kernels may be different according to the data sources. I will be checking your code in case you find smart things to do 😉 

PS: I will plagiarize some of your prose to improve my documentation 😁

**Comment by xoolive** [comment]
Also noteworthy is that in the end, the sigma depends on the aircraft: some transponders seem to be more noisy than others...

**Comment by espinielli** [comment]
fixed in fa452b0c16f39a8d7aaab2f91244f42319e7bd24
